### ISSUE TITLE
Investigate Sub-basin Errors

### ISSUE THREAD

**Issue by rajadain** [WikiWatershed/model-my-watershed#3457]
In this spreadsheet: [mapshed_water_quality_loads_Wissahickon.xlsx](https://github.com/WikiWatershed/model-my-watershed/files/7851500/mapshed_water_quality_loads_Wissahickon.xlsx), BME discovered that the numbers being sent to SRAT for component HUC-12s are larger than those shown in MMW when modeling just the HUC-12s themselves.

Investigate and correct.

From BME:

> Per our call yesterday, attached is a marked-up Excel file that shows results for the Wissahickon HUC10 basin near Philadelphia. To the left are results for the HUC10 without any sub-basin modeling, and to the right are results for the same basin after sub-basin modeling (and attenuation). As you can see, after this run several categories show an increase in loads instead of a decrease. This suggests to me that maybe estimated load reductions due to attenuation are being added to the "non-attenuated" loads rather than being subtracted. I haven't heard back from Mike yet about what he may have found, so there's no need for you to dig too deeply into this until I hear his response since we don't know whether the error is occurring on the API side before the results are sent back to MMW, or after. However, maybe what you could do in the meantime is too verify what data are being returned to MMW immediately after the Drexel sub-basin API is executed so that we can better understand what may be happening. Thanks.

> I've put the load results shown in the Manatawny JSON "request" file that you gave me into another Excel file (see attached) and compared them with the results given by MMW for each of the HUC12s comprising the Manatawny HUC10, and the results are different than what we got for the Wissahickon HUC10. As you recall, with the Wissahickon, the JSON request results for one HUC12 were OK for all values, but higher than expected for Hay/Pasture and Cropland for the other HUC12. However, for the Manatawny (which has 3 HUC12s), the results for one HUC12 were OK, but the Hay/Pasture loads for the other two HUC12s were lower than expected (and the Cropland results were OK). I'm not sure what to make of this, but hopefully you can sort it out. Let me know if we need to discuss this further. Thanks.

**Comment by rajadain** [comment]
Also check if we're using different point sources for sub-basin vs normal MapShed modeling.

**Comment by rajadain** [comment]
I added the following print statements and turned off geoprocessing caching:

```diff
diff --git a/src/mmw/apps/modeling/tasks.py b/src/mmw/apps/modeling/tasks.py
index 8fa654e4..877f7180 100644
--- a/src/mmw/apps/modeling/tasks.py
+++ b/src/mmw/apps/modeling/tasks.py
@@ -338,6 +338,8 @@ def run_gwlfe(model_input, inputmod_hash, watershed_id=None):
     directly we would have to replicate all that logic. Thus, it is easier to
     simply create a GMS file and have it read that.
     """
+    print(f'==> model input\n{json.dumps(model_input)}')
+
     output = to_gms_file(model_input)

     reader = Parser.GmsReader(output)
@@ -347,6 +349,8 @@ def run_gwlfe(model_input, inputmod_hash, watershed_id=None):
     result['inputmod_hash'] = inputmod_hash
     result['watershed_id'] = watershed_id

+    print(f'==> gwlfe result\n{json.dumps(result)}')
+
     return result
```

I then saved the files for Upper Wissahickon HUC-12, by running it solo, and then within the Wissahickon HUC-10 as subbasin:
- [upper-wissahickon-solo-model-input.json.txt](https://github.com/WikiWatershed/model-my-watershed/files/7952445/upper-wissahickon-solo-model-input.json.txt)
- [upper-wissahickon-solo-gwlfe-result.json.txt](https://github.com/WikiWatershed/model-my-watershed/files/7952446/upper-wissahickon-solo-gwlfe-result.json.txt)
- [upper-wissahickon-subbasin-model-input.json.txt](https://github.com/WikiWatershed/model-my-watershed/files/7952448/upper-wissahickon-subbasin-model-input.json.txt)
- [upper-wissahickon-subbasin-gwlfe-result.json.txt](https://github.com/WikiWatershed/model-my-watershed/files/7952450/upper-wissahickon-subbasin-gwlfe-result.json.txt)
 

Diffing the model inputs, we see:

```diff
--- upper-wissahickon-solo-model-input.json.txt	2022-01-27 11:47:09.000000000 -0500
+++ upper-wissahickon-subbasin-model-input.json.txt	2022-01-27 11:58:41.000000000 -0500
@@ -110,8 +110,8 @@
     "BasinDeadStorage": 0,
     "C":
     [
-        0.01,
-        0.24,
+        0.01574884872666669,
+        0.24287442436333323,
         0.002,
         0.01,
         0.08,
@@ -25496,6 +25496,5 @@
     "n85t": 0.71,
     "n85u": 0.82,
     "n85v": 0.71,
-    "n9": 0,
-    "watershed_id": null
+    "n9": 0
 }
\ No newline at end of file
```

So almost no difference, except for the absence of "watershed_id" (which I don't think is used within GWLF-E), and the additional resolution of those two `C` values.

However, the output has significant differences:

```diff
--- upper-wissahickon-solo-gwlfe-result.json.txt	2022-01-27 11:47:34.000000000 -0500
+++ upper-wissahickon-subbasin-gwlfe-result.json.txt	2022-01-27 11:58:55.000000000 -0500
@@ -3,16 +3,16 @@
     "Loads":
     [
         {
-            "Sediment": 103729.43233518643,
+            "Sediment": 163361.91383498593,
             "Source": "Hay/Pasture",
-            "TotalN": 767.0687679568882,
-            "TotalP": 302.1606574227851
+            "TotalN": 917.3009364949463,
+            "TotalP": 367.23836753630377
         },
         {
-            "Sediment": 64134.66912345614,
+            "Sediment": 64902.79518788438,
             "Source": "Cropland",
-            "TotalN": 280.7094725281395,
-            "TotalP": 86.18917796552937
+            "TotalN": 282.6446132680071,
+            "TotalP": 87.02744402614377
         },
         {
             "Sediment": 2376.831276623954,
@@ -98,31 +98,31 @@
     "SummaryLoads":
     [
         {
-            "Sediment": 5640840.72439673,
+            "Sediment": 5701241.3319609575,
             "Source": "Total Loads",
-            "TotalN": 29411.926124606856,
-            "TotalP": 29079.938826929192,
+            "TotalN": 29564.09343388478,
+            "TotalP": 29145.854803103324,
             "Unit": "kg"
         },
         {
-            "Sediment": 792.341937913913,
+            "Sediment": 800.8261225925605,
             "Source": "Loading Rates",
-            "TotalN": 4.131352697579342,
-            "TotalP": 4.084720028504494,
+            "TotalN": 4.152726912277331,
+            "TotalP": 4.093978930652787,
             "Unit": "kg/ha"
         },
         {
-            "Sediment": 142.8350767236035,
+            "Sediment": 144.36451636515824,
             "Source": "Mean Annual Concentration",
-            "TotalN": 0.7447568420834056,
-            "TotalP": 0.7363503946313509,
+            "TotalN": 0.7486099608572673,
+            "TotalP": 0.7380194921957305,
             "Unit": "mg/l"
         },
         {
-            "Sediment": 218.11744262186022,
+            "Sediment": 222.87026913543974,
             "Source": "Mean Low-Flow Concentration",
-            "TotalN": 1.3709254288856143,
-            "TotalP": 1.969558996741342,
+            "TotalN": 1.3828992293356697,
+            "TotalP": 1.9747458187264653,
             "Unit": "mg/l"
         }
     ],
@@ -260,5 +260,5 @@
             "AvWithdrawal": 0.0
         }
     ],
-    "watershed_id": null
+    "watershed_id": "020402030901"
 }
\ No newline at end of file
```

Going to investigate why that is happening.

**Comment by rajadain** [comment]
I added some manipulation in the code to hard-set those values to be without the extended digits:

```diff
diff --git a/src/mmw/apps/modeling/tasks.py b/src/mmw/apps/modeling/tasks.py
index 8fa654e4..e40d4689 100644
--- a/src/mmw/apps/modeling/tasks.py
+++ b/src/mmw/apps/modeling/tasks.py
@@ -338,6 +338,11 @@ def run_gwlfe(model_input, inputmod_hash, watershed_id=None):
     directly we would have to replicate all that logic. Thus, it is easier to
     simply create a GMS file and have it read that.
     """
+    model_input['C'][0] = 0.01
+    model_input['C'][1] = 0.24
+
+    print(f'==> model input\n{json.dumps(model_input)}')
+
     output = to_gms_file(model_input)

     reader = Parser.GmsReader(output)
@@ -347,6 +352,8 @@ def run_gwlfe(model_input, inputmod_hash, watershed_id=None):
     result['inputmod_hash'] = inputmod_hash
     result['watershed_id'] = watershed_id

+    print(f'==> gwlfe result\n{json.dumps(result)}')
+
     return result
```

And using that generated manipulated versions of the subbasin model input and gwlfe result:

- [upper-wissahickon-subbasin-model-input-manipulated.json.txt](https://github.com/WikiWatershed/model-my-watershed/files/7961275/upper-wissahickon-subbasin-model-input-manipulated.json.txt)
- [upper-wissahickon-subbasin-gwlfe-result-manipulated.json.txt](https://github.com/WikiWatershed/model-my-watershed/files/7961277/upper-wissahickon-subbasin-gwlfe-result-manipulated.json.txt)
 
Upon diffing these with the originals, we see that there is no longer a difference in `C` value, thanks to the manipulation:

```diff
--- upper-wissahickon-solo-model-input.json.txt	2022-01-27 11:47:09.000000000 -0500
+++ upper-wissahickon-subbasin-model-input-manipulated.json.txt	2022-01-28 12:41:54.000000000 -0500
@@ -25496,6 +25496,5 @@
     "n85t": 0.71,
     "n85u": 0.82,
     "n85v": 0.71,
-    "n9": 0,
-    "watershed_id": null
+    "n9": 0
 }
\ No newline at end of file
```

and the output is identical as well:

```diff
--- upper-wissahickon-solo-gwlfe-result.json.txt	2022-01-27 11:47:34.000000000 -0500
+++ upper-wissahickon-subbasin-gwlfe-result-manipulated.json.txt	2022-01-28 12:42:46.000000000 -0500
@@ -260,5 +260,5 @@
             "AvWithdrawal": 0.0
         }
     ],
-    "watershed_id": null
+    "watershed_id": "020402030901"
 }
\ No newline at end of file
```

So it is the additional decimal digits causing this.

**Comment by rajadain** [comment]
This seems to be the offending line:

https://github.com/WikiWatershed/model-my-watershed/blob/12472c8af2126bcf582faa2f72030b4daa3118ae/src/mmw/apps/modeling/mapshed/tasks.py#L52-L53

I added the following statements before and after the value of `C` was modified:

```diff
diff --git a/src/mmw/apps/modeling/mapshed/tasks.py b/src/mmw/apps/modeling/mapshed/tasks.py
index cd8c6b42..b13091cf 100644
--- a/src/mmw/apps/modeling/mapshed/tasks.py
+++ b/src/mmw/apps/modeling/mapshed/tasks.py
@@ -74,9 +74,11 @@ def collect_data(geop_results, geojson, watershed_id=None, weather=None,
     z['WxYrs'] = z['WxYrEnd'] - z['WxYrBeg'] + 1

     # Data from the County Animals dataset
+    print(f'==> z.C BEFORE\n{z["C"]}')
     ag_lscp = ag_ls_c_p(geom)
     z['C'][0] = ag_lscp.hp_c
     z['C'][1] = ag_lscp.crop_c
+    print(f'==> z.C AFTER\n{z["C"]}')

     livestock_aeu, poultry_aeu, population = animal_energy_units(geom)
     z['AEU'] = livestock_aeu / (area * ACRES_PER_SQM)
```

And then in the Celery output saw the following:

```
[2022-01-28 15:42:16,076: INFO/MainProcess] Task apps.modeling.mapshed.tasks.collect_data[6ac7f14a-764c-446e-9c3f-695252570a04] received
[2022-01-28 15:42:16,168: WARNING/ForkPoolWorker-1] ==> z.C BEFORE
[0.03, 0.42, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
[2022-01-28 15:42:16,205: WARNING/ForkPoolWorker-1] ==> z.C AFTER
[0.013263110134957127, 0.24163155506747866, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
```
```
[2022-01-28 15:42:26,205: INFO/MainProcess] Task apps.modeling.mapshed.tasks.collect_subbasin[4d4dae9b-038e-49cc-ac9e-b8a1dc663328] received
[2022-01-28 15:42:26,479: WARNING/ForkPoolWorker-1] ==> z.C BEFORE
[0.013263110134957127, 0.24163155506747866, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
[2022-01-28 15:42:26,497: WARNING/ForkPoolWorker-1] ==> z.C AFTER
[0.01, 0.24, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
[2022-01-28 15:42:26,602: WARNING/ForkPoolWorker-1] ==> z.C BEFORE
[0.01, 0.24, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
[2022-01-28 15:42:26,630: WARNING/ForkPoolWorker-1] ==> z.C AFTER
[0.01574884872666669, 0.24287442436333323, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
```

The first block is from the Wissahickon HUC-10 Mapshed run, and the second block is from the same shape's subbasin run, which includes two executions, one for the Upper Wissahickon HUC-12, and one for the Lower Wissahickon HUC-12.

We see the in the first BEFORE, the true defaults are used, as defined here:

https://github.com/WikiWatershed/model-my-watershed/blob/12472c8af2126bcf582faa2f72030b4daa3118ae/src/mmw/mmw/settings/gwlfe_settings.py#L42-L43

However, the second BEFORE uses the values of the first AFTER, and the third BEFORE those of the second AFTER. This pollution is happening because dict.copy creates a 1-level deep copy of the underlying dictionary, not a true deep copy.

Going to switch to deepcopy and see if that helps. This is similar to the bug fixed in 8a55f88f3d3c21b09ddf14046889a3838d563174 .

**Comment by rajadain** [comment]
It does indeed. The following diff:

```diff
diff --git a/src/mmw/apps/modeling/mapshed/tasks.py b/src/mmw/apps/modeling/mapshed/tasks.py
index cd8c6b42..7363e473 100644
--- a/src/mmw/apps/modeling/mapshed/tasks.py
+++ b/src/mmw/apps/modeling/mapshed/tasks.py
@@ -1,4 +1,5 @@
 # -*- coding: utf-8 -*-
+from copy import deepcopy
 from celery import shared_task
 from django.conf import settings

@@ -50,7 +51,7 @@ def collect_data(geop_results, geojson, watershed_id=None, weather=None,
     area = geom.transform(5070, clone=True).area  # Square Meters

     # Data Model is called z by convention
-    z = settings.GWLFE_DEFAULTS.copy()
+    z = deepcopy(settings.GWLFE_DEFAULTS)

     z['watershed_id'] = watershed_id
```

Makes the output use the correct BEFORE values in every case:

```
[2022-01-28 15:51:05,107: INFO/MainProcess] Task apps.modeling.mapshed.tasks.collect_data[7fc1d263-dd6f-4883-a552-cb5951369347] received
[2022-01-28 15:51:05,206: WARNING/ForkPoolWorker-2] ==> z.C BEFORE
[0.03, 0.42, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
[2022-01-28 15:51:05,242: WARNING/ForkPoolWorker-2] ==> z.C AFTER
[0.013263110134957127, 0.24163155506747866, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
```
```
[2022-01-28 15:51:12,509: INFO/MainProcess] Task apps.modeling.mapshed.tasks.collect_subbasin[a094710e-f321-4694-83b6-9ef76a9e66d5] received
[2022-01-28 15:51:12,792: WARNING/ForkPoolWorker-1] ==> z.C BEFORE
[0.03, 0.42, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
[2022-01-28 15:51:12,812: WARNING/ForkPoolWorker-1] ==> z.C AFTER
[0.01, 0.24, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
[2022-01-28 15:51:12,919: WARNING/ForkPoolWorker-1] ==> z.C BEFORE
[0.03, 0.42, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
[2022-01-28 15:51:12,946: WARNING/ForkPoolWorker-1] ==> z.C AFTER
[0.01574884872666669, 0.24287442436333323, 0.002, 0.01, 0.08, 0.03, 0.04, 0.001, 0.01, 0.8, 0, 0, 0, 0, 0, 0]
```
